### PR TITLE
Fix TarArchiveLoader to skip `open` for TarFile stream

### LINKED ISSUE #673: .open_files(mode="b").decompress().load_from_tar() throws error
### 🐛 Describe the bug

repo - https://github.com/MatthewCaseres/torchdata-slowpipe-repro
```py
from torchdata.datapipes.iter import FileLister

dp = FileLister("caltech-101").open_files(mode="b").decompress().load_from_tar()
for x in dp:
    print(x)
    break
```
throws error - 
```
/home/ubuntu/.local/lib/python3.9/site-packages/torchdata/datapipes/iter/util/tararchiveloader.py:72: UserWarning: Unable to extract files from corrupted tarfile stream caltech-101/101_ObjectCategories.tar.gz due to: 'TarFile' object has no attribute 'tell', abort!
  warnings.warn(f"Unable to extract files from corrupted tarfile stream {pathname} due to: {e}, abort!")
Traceback (most recent call last):
  File "/home/ubuntu/torchdata-slowpipe-repro/script.py", line 4, in <module>
    for x in dp:
  File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/data/datapipes/_typing.py", line 514, in wrap_generator
    response = gen.send(None)
  File "/home/ubuntu/.local/lib/python3.9/site-packages/torchdata/datapipes/iter/util/tararchiveloader.py", line 73, in __iter__
    raise e
  File "/home/ubuntu/.local/lib/python3.9/site-packages/torchdata/datapipes/iter/util/tararchiveloader.py", line 61, in __iter__
    tar = tarfile.open(fileobj=cast(Optional[IO[bytes]], data_stream), mode=self.mode)
  File "/opt/conda/envs/pytorch/lib/python3.9/tarfile.py", line 1618, in open
    saved_pos = fileobj.tell()
  File "/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/data/datapipes/utils/common.py", line 187, in __getattr__
    return getattr(file_obj, name)
AttributeError: 'TarFile' object has no attribute 'tell'
This exception is thrown by __iter__ of TarArchiveLoaderIterDataPipe(datapipe=DecompressorIterDataPipe, length=-1, mode='r:*')
```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.13.0-1031-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4
GPU 2: Tesla T4
GPU 3: Tesla T4

Nvidia driver version: 510.73.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.12.0
[pip3] torch-model-archiver==0.5.3b20220226
[pip3] torch-workflow-archiver==0.2.4b20220513
[pip3] torchaudio==0.12.0
[pip3] torchdata==0.4.0
[pip3] torchserve==0.6.0b20220513
[pip3] torchtext==0.13.0
[pip3] torchvision==0.13.0
[conda] blas                      2.115                       mkl    conda-forge
[conda] blas-devel                3.9.0            15_linux64_mkl    conda-forge
[conda] captum                    0.5.0                         0    pytorch
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] libblas                   3.9.0            15_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            15_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            15_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            15_linux64_mkl    conda-forge
[conda] magma-cuda116             2.6.1                         0    pytorch
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.12.0          py3.9_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-model-archiver      0.5.3                    py39_0    pytorch
[conda] torch-workflow-archiver   0.2.4                    py39_0    pytorch
[conda] torchaudio                0.12.0               py39_cu116    pytorch
[conda] torchserve                0.6.0                    py39_0    pytorch
[conda] torchtext                 0.13.0                     py39    pytorch
[conda] torchvision               0.13.0               py39_cu116    pytorch
```

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #673

If `data_stream` is either a `tafile.TarFile` object (wrapped by `StreamWrapper`), `TarArchiveLoader` should directly load from `data_stream` without `open` it again.

### CODE REVIEW

File: torchdata/datapipes/iter/util/tararchiveloader.py [hunk 1]
```diff
@@ -57,9 +57,18 @@ def __iter__(self) -> Iterator[Tuple[str, BufferedIOBase]]:
             validate_pathname_binary_tuple(data)
             pathname, data_stream = data
             try:
-                reading_mode = self.mode if data_stream.seekable() else self.mode.replace(":", "|")
-                # typing.cast is used here to silence mypy's type checker
-                tar = tarfile.open(fileobj=cast(Optional[IO[bytes]], data_stream), mode=reading_mode)
+                if isinstance(data_stream, tarfile.TarFile):
+                    tar = data_stream
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/tararchiveloader.py:61]: super nit: can we add a small test for this case? (only if it is easy to add)

**Review by NivekT — APPROVED** [review]
LGTM! One small nit

File: torchdata/datapipes/iter/util/tararchiveloader.py (hunk 1 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/tararchiveloader.py:61]: Make sense. I will add one via `map`.

Comment (ejguan) [reply torchdata/datapipes/iter/util/tararchiveloader.py:61]: After digging deeper on the implementation, I will remove this condition as `validate_pathname_binary_tuple ` won't allow the stream being `tarfile.TarFile` directly. It must be wrapped by `StreamWrapper`.
https://github.com/pytorch/data/blob/6b8c5780d43d80854f340df60825a7e624c1ec81/torchdata/datapipes/iter/util/tararchiveloader.py#L57

**Pull request closed without merging**